tests/robustness/multiclient_test/framework/snapshotter.go

```diff
@@ -0,0 +1,192 @@
+// +build darwin,amd64 linux,amd64
+
+package framework
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"log"
+	"os/exec"
+	"strconv"
+	"sync"
+
+	"github.com/kopia/kopia/tests/robustness"
+)
+
+// MultiClientSnapshotter manages a set of client Snapshotter instances and
+// implements the Snapshotter interface itself. Snapshotter methods must be
+// provided with a client-wrapped context so the MultiClientSnapshotter can
+// delegate to a specific client Snapshotter.
+type MultiClientSnapshotter struct {
+	baseDirPath string
+	server      Server
+
+	// Map of client ID to ClientSnapshotter and associated lock
+	clients map[string]ClientSnapshotter
+	mu      sync.RWMutex
+
+	// Function used to generate new ClientSnapshotters
+	newClientSnapshotter newClientFn
+}
+
+// MultiClientSnapshotter implements robustness.Snapshotter.
+var _ robustness.Snapshotter = (*MultiClientSnapshotter)(nil)
+
+type newClientFn func(string) (ClientSnapshotter, error)
+
+// NewMultiClientSnapshotter returns a MultiClientSnapshotter that is
+// responsible for delegating Snapshotter method calls to a specific client's
+// Snapshotter instance. ConnectOrCreateRepo must be invoked to start the server.
+func NewMultiClientSnapshotter(baseDirPath string, f newClientFn, s Server) (*MultiClientSnapshotter, error) {
+	return &MultiClientSnapshotter{
```

Comment:
Something like this for initializing the server?


```suggestion
	s, err := snapmeta.NewSnapshotter(baseDirPath)
	if err != nil {
		log.Println("Error creating kopia Snapshotter:", err)
		return nil, err
	}

	if sErr := s.ConnectOrCreateRepo(dataRepoPath); err != nil {
		return nil, sErr
	}

	_, _, err = s.Run("policy", "set", "--global", "--keep-latest", strconv.Itoa(1<<31-1), "--compression", "s2-default")
	if err != nil {
		return nil, err
	}
	return &MultiClientSnapshotter{
```

Reply:
I like the construction of the real Snapshotter here. I would not bundle `ConnectOrCreateRepo` method here: it is not "constructor" material. Your original logic was fine.

Reply:
Thanks. So based on discussions so far:
- `TestHarness.getSnapshotter` will call `NewMultiClientSnapshotter` and then `MultiClientSnapshotter.ConnectOrCreateRepo`.
- `NewMultiClientSnapshotter` will construct a new server snapshotter (uninitialized) using `snapmeta.NewSnapshotter` and returns new MultiClientSnapshotter with that server.
- `MultiClientSnapshotter.ConnectOrCreateRepo` remains unchanged. It calls `ConnectOrCreateRepo` on the server and runs policy setup on the repo.

Does that sound correct?